Fix bug in atomicAdd for int16_t

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29376 [WIP] Add type promotion support in sparse tensors
* **#29231 Fix bug in atomicAdd for int16_t**

Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```

Differential Revision: [D18575666](https://our.internmc.facebook.com/intern/diff/D18575666)

nairbv committed:
Fix bug in atomicAdd for int16_t

[ghstack-poisoned]

gchanan commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -40,7 +40,8 @@ struct AtomicAddIntegerImpl<T, 2> {
       assumed = old;
-      sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);
+      sum = (size_t)address & 2 ? old >> 16 : old & 0xffff;
+      sum = THCNumerics<T>::add(sum, val);
```
wait, how are the other cases not broken then?

nairbv replied:
it was chance that that half fixed the issue. After adding more unit tests, found that a reinterpret_cast is also required, and that int8 was also broken.

nairbv committed:
Update on "[WIP] Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153


[ghstack-poisoned]

zou3519 commented:
What is the bug? Is the original calculation just flat-out wrong? This information would be nice to have as well in the commit message / PR body

zou3519 commented on test/test_torch.py:
```diff
@@ -13367,6 +13367,35 @@ def test_cat_big(self, device):
 
+    @onlyCUDA
+    def test_atomic_add(self, device):
+        # todo: we should probably expose a dtype.is_signed()
```
Should file an issue for TODOs so we don't forget about them and link the issue from the TODO

zou3519 commented on test/test_torch.py:
```diff
@@ -13367,6 +13367,35 @@ def test_cat_big(self, device):
+            else:
+                tensor = torch.randint(0, 10, size, dtype=dtype, device=device)
+
+            # index_add calls atomicAdd on cuda.
```
It seems like we're relying on `index_add` calling atomicAdd. It might be clearer to introduce a C++ API test that only calls atomicAdd so that we're still testing this behavior if someone changes the implementation of `index_add`

nairbv replied:
I'll check with you tomorrow. Last I looked it didn't seem like we had the infrastructure setup to build and run tests on cu files. If you know an easy way to do that or where they'd go maybe point me to it, otherwise I think setting that up could be a separate task.

zou3519 replied:
Some of our cpp tests definitely run on CUDA: https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_fuser.cpp

zou3519 commented on test/test_torch.py:
```diff
@@ -13367,6 +13367,35 @@ def test_cat_big(self, device):
+            added = empty.index_add(0, torch.arange(0, size[0], dtype=torch.long, device=device), tensor)
+            self.assertEqual(added.sum(), tensor.sum())
+
+            # sparse.to_dense() on non-coalesced tensors calls index_add, which was where
```
Perhaps this should be a different test case (and should be placed inside `test_sparse`)?

nairbv replied:
it's probably not really necessary after I add the next stacked diff anyway. it was just my original test before I drilled deeper.

nairbv commented:
> What is the bug? Is the original calculation just flat-out wrong? This information would be nice to have as well in the commit message / PR body

it's in the linked issue, but I've now updated the PR to include the the repro here too.

zou3519 commented on test/test_torch.py:
```diff
@@ -13367,6 +13367,35 @@ def test_cat_big(self, device):
+        for dtype in torch.testing.get_all_math_dtypes(device):
+            size = [5, 5]
+            if dtype.is_floating_point:
+                tensor = torch.rand(size, dtype=dtype, device=device)
```
This line does nothing, right? All dtypes are either signed or unsigned and the value here is overwritten in one of the two branches that comes next.

zou3519 reviewed (commented):
(not a full review) I had some comments on the testing. Am trying to understanding what the bug was and what the fix is at the moment.

zou3519 commented:
> > What is the bug? Is the original calculation just flat-out wrong? This information would be nice to have as well in the commit message / PR body
> 
> it's in the linked issue, but I've now updated the PR to include the the repro here too. Yes the original code was just wrong, returned zeros.

What I am looking for is an English description of what the bug is and how this PR fixes it

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -21,7 +21,9 @@ struct AtomicAddIntegerImpl<T, 1> {
-      sum = val + T((old >> shift) & 0xff);
+      T tmp = THCNumerics<T>::add(val, T((old >> shift) & 0xff));
+      // reinterpret raw bytes to avoid miss-interpretation of sign
+      sum = *reinterpret_cast<uint16_t*>(&tmp);
```
Should this be `uint8_t`?

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -40,7 +42,10 @@ struct AtomicAddIntegerImpl<T, 2> {
     do {
       assumed = old;
-      sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);
+      sum = (size_t)address & 2 ? old >> 16 : old & 0xffff;
```
Why are we casting `address` to `size_t`, but we're using `uint32_t` for it above? Does CUDA address memory using 32 bits?

nairbv replied:
I was trying to change this as little as possible since it's a bit confusing, but, my understanding of this code is that we can only atomically compare-and-swap 32 bits of memory at a time, and only from a "round" memory addresses. What we're checking with `(size_t)address & 2` is if the *address* is odd/even (if it points to directly addressable memory or if we have to shift to CAS the value we want). I think `size_t` is just always going to be the size of the address, regardless of if it's 32 or 64 bit. The size of the address is unrelated to the size of the value stored at that address, and the value we actually care about is half (or a quarter) of the size of the smallest number of bits we can actually directly CAS.

The uint32_t is just an arbitrary 32 bit type to hold the smallest number of bits we can atomically compare-and-swap from the address. We want to avoid doing math on that uint32, since it's really holding one or more values in the format of `T`.

zou3519 replied:
Doing `(size_t)address & 2` is functionally equivalent to `address & 2` but potentially slower due to doing the computation in int64_t, as far as I can tell. Does it look different to you? If not we should take this chance to clean up the code anyways.

nairbv replied:
I think you're commenting on an older version of this? I made some changes Friday

zou3519 commented on test/test_torch.py:
```diff
@@ -13395,6 +13395,35 @@ def test_cat_big(self, device):
+            if is_signed(dtype):
+                tensor = torch.randint(-5, 15, size, dtype=dtype, device=device)
+            else:
+                tensor = torch.randint(0, 10, size, dtype=dtype, device=device)
```
The atomicAdd bug looks like it is triggered in a very specific situation, when the sum being stored back into a memory address is signed. Would it be worth testing that case specifically?

nairbv replied:
It needs to handle sign properly, but it wasn't just that: the original code just always returned zero. In intermediate versions I thought I fixed the issue, but I ran into issues that only occurred at specific alignments of values within the 32 bit words. In practice I don't feel like I got good test signal when I wrote specific narrowly defined tests for particular cases, and the randint for the signed case should always get some negatives.

zou3519 replied:
Could you give some sample inputs to index_add that trigger the bug? I want to know roughly the order of magnitude of the probability is that the randomly generated input does not catch the bug; we want to make our tests robust to changes in seed.

nairbv replied:
In the example from the initial issue we actually got all zeros, so almost any random value would catch that. There are lots of possible ways this can be wrong though that were missed when I just tried to validate narrowly for particular values, and random tensors capture that more thoroughly.

zou3519 replied:
Could you walk through an example input and how it gets to all zeros? There's a lot of bit shifting going on and I'd like to make sure I understand how the implementation works

nairbv replied:
I mean, nothing gets added at all. I'm not actually sure if there's a case that doesn't trigger this bug besides maybe adding a tensor of zeros. Here's an example for a single value with the expected result for int64 and the zero result for int16, it's basically a no-op:

```
# expected 0 + 1 == 1
>>> dtype=torch.int64
>>> torch.zeros([1], device='cuda', dtype=dtype).index_add(0, torch.tensor([0], device='cuda'), torch.ones([1], dtype=dtype, device='cuda'))
tensor([1], device='cuda:0')
>>> dtype=torch.int16
# unexpected 0 + 1 == 0
>>> torch.zeros([1], device='cuda', dtype=dtype).index_add(0, torch.tensor([0], device='cuda'), torch.ones([1], dtype=dtype, device='cuda'))
tensor([0], device='cuda:0', dtype=torch.int16)
```

It would be easy to add this example that motivated the change, but it's just so trivial that it doesn't add coverage. The test with a full multi-dimensional tensor and a variety of values can capture potential corner cases that this narrow test wouldn't.

zou3519 replied:
I'm not complaining about the test case coverage, I'm still trying to understand why the previous code is wrong. Because the fix was to change from doing unsigned math to signed math, there should be a non-trivial set of positive values for which the math is correct, but this doesn't seem to be the case from your explanation.

To help bridge my gap in understanding, it would be really helpful if you could run through an end-to-end example about how almost all inputs give zeros.

nairbv replied:
I'm not sure it's worth trying to dig too deeply into all the ways the existing code was incorrect. It didn't work in any situation I know of and was never tested. There's more wrong with it than just the sign issue.

One thing that's wrong is that in `sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);`, ... the `val + (size_t)address &2` as an expression has precedence. If val is non-zero, we'll always call `T(old>>16)` regardless of what the address was. Just fixing that, we still have the sign issue, but I'm not sure that completely covers the problems here.

There are so many ways untested code can be wrong, it's hard to be certain of all of them without a lot of investigation and additional testing.

nairbv commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -21,7 +21,9 @@ struct AtomicAddIntegerImpl<T, 1> {
     do {
       assumed = old;
-      sum = val + T((old >> shift) & 0xff);
+      T tmp = THCNumerics<T>::add(val, T((old >> shift) & 0xff));
```
I believe that's right, though I guess maybe it could also be bool? I think this should be able to handle any 1 byte type supported by THCNumerics.

nairbv commented:
for reference it looks like the original code may have been adapted from:

https://github.com/rapidsai/cudf/blob/8ddb744a716aa8fe736284272b065cf5b7085065/cpp/src/reductions/reduction_operators.cuh#L20-L43
or
https://github.com/rapidsai/cudf/issues/1179

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]

gchanan commented on test/test_torch.py:
```diff
@@ -13395,6 +13395,25 @@ def test_cat_big(self, device):
         result = torch.cat(concat_list)
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    @onlyCUDA
```
should probably run on all devices -- there's nothing specific about CUDA here (just that the bug happened to be in CUDA).

zou3519 replied:
We should rename the test to reference "index_add" if we remove the @onlycuda, the TH code doesn't use "atomic add" (it does the computation serially)

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -12,36 +12,42 @@ struct AtomicAddIntegerImpl;
-      old = atomicCAS(address_as_ui, assumed, old);
+      sum = (old >> shift) & 0xff;
+      T tmp = THCNumerics<T>::add(val, sum);
+      sum = *reinterpret_cast<uint8_t*>(&tmp);
```
Sorry, I lost my train of thought from last week -- why is this necessary again? What happens if we don't have this? An example would help here.

zou3519 replied:
This seems equivalent to `sum = THCNumerics<T>::add(val, sum)`

nairbv replied:
if I do:
```
   int8_t a = -5;
   uint32_t b = *reinterpret_cast<uint32_t*>(&a);
   uint32_t c = (uint32_t)a;
```
b != c

if it was just the change in sign they'd be the same, but since it changes sign and size the interpreted value changes. 

as an example, -1 to unsigned is always ~0, so if the destination is bigger it fills in 1's in bits we didn't want to change. we just want the raw bytes of the source memory space.

zou3519 replied:
Got it, thank you for the example, that helped a lot.

The reinterpret_cast might violate strict aliasing and could lead to UB: see https://xania.org/200712/cpp-strict-aliasing and https://stackoverflow.com/questions/14623266/why-cant-i-reinterpret-cast-uint-to-int . Could we switch this to one of the solutions in the latter link?

zou3519 replied:
Also, could you include a comment in the code that summarizes what you've explained here?

nairbv replied:
Are we sure the strict aliasing problem applies here? I'd rather not complicate the code if it's not actually necessary to add the structs it would take.

What I'm seeing is that if we have two pointers of different types, during optimization the compiler can assume that they are not aliases/pointers to the same location in memory when making optimizations.

as an example, if we had:
```
f(int* x, int * y) {
*x=0
*y=1
return *x;
```
`*y=1` _could_ overwrite the value in x, because the compiler doesn't know if y is a pointer to the same location as x. If x and y are pointers to different types, the return statement can be compiled as `return 0` because the compiler is expected to assume `*y=1` cannot change the return value (can't be an alias to x).

In this code, we're just copying the value out of a pointed to address and interpreting it differently. we don't maintain multiple pointers/aliases of different types (sum isn't a pointer), so I'm not seeing where we can actually get UB. There's a transient pointer, but it's just read from and thrown away. For the compiler to pretend that temporary address isn't an alias doesn't seem like it would be problematic.

Also it looks like we compile with -fno-strict-aliasing

zou3519 replied:
I'm not sure so I am raising the question. The link here: https://stackoverflow.com/questions/14623266/why-cant-i-reinterpret-cast-uint-to-int looks like exactly what we're doing (scroll down to the section on reinterpret_cast)

nairbv replied:
I think I have a simpler change you'll like: if I initially cast to an unsigned type of the same size, there's no extra space to pad with 1's in the case of a negative signed value. This seems to work with a regular cast without needing to reinterpret bytes, use a union, or memcpy.

zou3519 replied:
Yeah, this is much better now, thank you for finding it

nairbv commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -12,36 +12,42 @@ struct AtomicAddIntegerImpl;
-      sum = val + T((old >> shift) & 0xff);
-      old = (old & ~(0x000000ff << shift)) | (sum << shift);
-      old = atomicCAS(address_as_ui, assumed, old);
+      sum = (old >> shift) & 0xff;
```
we do add into it below though. if it's `old_byte` then we might need more variables, since it's not old anymore below

zou3519 replied:
"(old >> shift) & 0xff" isn't a sum -- calling it a sum made me take a double take and check if the the shift and bitwise and are actually magically computing a sum

nairbv replied:
above we're putting the initial value in sum, then below we add val to it.

it's effectively just:
sum = old
tmp = sum + val
sum = tmp
but with a bunch of casting and bit shifting

zou3519 replied:
To give an analogy, the code is like:
pi = 220 / 10
tmp = pi / 7
pi = tmp

When reading this, the reader sees "pi = 220 / 10" and thinks, "hey, is this some different way to compute pi?". Upon computing it in actuality, they discover that "no, this number isn't pi -- maybe I'm doing it wrong". After a while, they realize that the variable is being written to in-place and that pi is actually pi later on.

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -13415,6 +13415,26 @@ def test_cat_big(self, device):
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    def test_index_add(self, device):
+        # test coverage for issue with atomic add:
```
I don't understand what "test coverage for issue with atomic add" means. How about something like "we don't implement dtype.is_signed in core, see issue"

nairbv replied:
oh, I meant to comment with a link to the relevant ticket, because I was changing the name and want to make it clear that what we're hitting with this test is atomic add

zou3519 commented on test/test_torch.py:
```diff
@@ -13415,6 +13415,26 @@ def test_cat_big(self, device):
+        # test coverage for issue with atomic add:
+        # https://github.com/pytorch/pytorch/issues/29475
+        def is_signed(dtype):
+            return torch.is_signed(torch.tensor(0, dtype=dtype))
```
nit: you can actually create a very empty tensor with `torch.empty([], dtype=dtype)`. It's marginally faster than creating a tensor with data because it doesn't have to allocate any data.

zou3519 commented on test/test_torch.py:
```diff
@@ -13415,6 +13415,26 @@ def test_cat_big(self, device):
+        def is_signed(dtype):
+            return torch.is_signed(torch.tensor(0, dtype=dtype))
+        for dtype in torch.testing.get_all_math_dtypes(device):
+            # print(dtype)
```
Forgot to delete comment?. One could toss the `dtype` into the self.assertEquals message if we really wanted to know what we're testing.

nairbv replied:
I'll remove it, but I do sometimes like leaving easy-to-uncomment prints in tests sometimes, especially when constructing the info to print is more complicated.
In this case I don't always remember the order of the dtypes, so in development it's interesting to see which dtypes passed before a failure. changing just the assert doesn't fix that.

zou3519 replied:
In that case, the solution to the problem should be to add one test per dtype

zou3519 commented on test/test_torch.py:
```diff
@@ -13415,6 +13415,26 @@ def test_cat_big(self, device):
         result = torch.cat(concat_list)
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    def test_index_add(self, device):
```
We have a test_index_add already: https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L2603

Could we put these two tests closer together or maybe consolidate them? Does running that test on CUDA catch this bug (it seems cpu-only)?

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -12,36 +12,46 @@ struct AtomicAddIntegerImpl;
-      sum = val + T((old >> shift) & 0xff);
-      old = (old & ~(0x000000ff << shift)) | (sum << shift);
-      old = atomicCAS(address_as_ui, assumed, old);
+      old_bytes = (old >> shift) & 0xff;
```
nit: This is one byte, not "bytes".

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -12,36 +12,46 @@ struct AtomicAddIntegerImpl;
   inline __device__ void operator()(T *address, T val) {
-    uint32_t * address_as_ui =
-        (uint32_t *) (address - ((size_t)address & 3));
+    size_t offset = (size_t)address & 3;
```
There's some rampant casting going on in this function. I know that we don't follow any style guides whatsoever, but in general, c-style casts are not very "c++": https://google.github.io/styleguide/cppguide.html#Casting . What are your thoughts on this subject?

nairbv replied:
I think I prefer the explicit style in those examples. I'd like to see us move towards having more formal style guides and I think there are good reasons for C++ style to differ from C. 

One concern though is I'm fixing a small bug in two of these addIntegerImpl's, but there are a number of these which all look very similar. Things that make this particular impl easier to read could also make it inconsistent with the others. If I go too far in changing the style I might prefer to do that in a separate PR and do all impls.

zou3519 commented on aten/src/THC/THCAtomics.cuh:
```diff
@@ -12,36 +12,46 @@ struct AtomicAddIntegerImpl;
-      old = atomicCAS(address_as_ui, assumed, old);
+      old_bytes = (old >> shift) & 0xff;
+      T tmp = THCNumerics<T>::add(val, old_bytes);
+      // maintain same size in initial cast to avoid padding negative values with 1s
```
As the code reader I'd like a little more information. Something like:
"If we did something like uint32_t newval = tmp (which is -1), then newval = 0xffffffff, but we really just wanted to store one byte (0x000000ff)."

nairbv replied:
updated a bit, also made the cast in-line instead of extra tmp values.

zou3519 reviewed (commented):
Code looks correct. Had some comments about the tests

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -2620,6 +2620,26 @@ def test_index_add(self):
+                else:
+                    tensor = torch.randint(0, 10, size, dtype=dtype, device=device)
+                # index_add calls atomicAdd on cuda.
+                empty = torch.zeros(size, dtype=dtype, device=device)
```
nit: this isn't an "empty" tensor, it is zero-filled so it is better to call it something like "zeros"

zou3519 commented on test/test_torch.py:
```diff
@@ -2620,6 +2620,26 @@ def test_index_add(self):
+    # add coverage for issue with atomic add that appeared only for 
+    # specific dtypes on cuda:
+    # https://github.com/pytorch/pytorch/issues/29153
+    def test_dtypes_for_index_add(self):
```
I wouldn't call this "test_dtypes_for_index_add". The other test doesn't test dtypes, but it really should (on the other hand, `test_index_add` shouldn't test CUDA because then it would run for too long).

What you're doing here is checking that an index_add between zeros and a tensor gives back the original tensor. How about naming this something like `test_index_add_trivial` or something

nairbv replied:
I don't like "trivial" because it doesn't capture what's actually tested. The main purpose of this test is to capture if index_add doesn't work for some dtypes due to special dtype handling code. I could do another PR at some point to merge the two though.

zou3519 replied:
Again, I feel really weird having both tests. In terms of design, it makes sense for all tests in test_torch, unless otherwise specified, to test all dtypes; something named "test_index_add" should really test everything unless there are multiple tests that have a prefix named "test_index_add" that have differentiating suffixes. I don't want to block this fix on that but it would be great if you could send a followup PR to clean the tests up.

nairbv replied:
I'm fine to eventually merge the two. I agree all of these tests should eventually be running for all dtypes.

zou3519 commented on test/test_torch.py:
```diff
@@ -2620,6 +2620,26 @@ def test_index_add(self):
             dest2[idx[i]] = dest2[idx[i]] + src[i]
         self.assertEqual(dest, dest2)
 
+    # add coverage for issue with atomic add that appeared only for 
```
It would be nice (doesn't have to happen in this PR) to change `test_index_add` to work with dtypes as well to add more test coverage.

zou3519 reviewed (approved):
Code looks fine, had some nits with the testing

nairbv committed:
Update on "Fix bug in atomicAdd for int16_t"


Fixes: https://github.com/pytorch/pytorch/issues/29153

Bug is that atomicAdd doesn't correctly add values for some dtypes due to incorrect casting. Was returning zeros.

Incorrect behavior before this PR:

```
In [23]: sparse=torch.sparse_coo_tensor(indices=torch.tensor([[0,0],[1,1]]), values=torch.tensor([5, 6], dtype=torch.int16), size=(2,2), device='cuda', dtype=torch.int16 )

In [24]: sparse
Out[24]:
tensor(indices=tensor([[0, 0],
                       [1, 1]]),
       values=tensor([5, 6]),
       device='cuda:0', size=(2, 2), nnz=2, dtype=torch.int16,
       layout=torch.sparse_coo)

In [25]: sparse.coalesce()
Out[25]:
tensor(indices=tensor([[0],
                       [1]]),
       values=tensor([11]),
       device='cuda:0', size=(2, 2), nnz=1, dtype=torch.int16,
       layout=torch.sparse_coo)

In [26]: sparse.to_dense()
Out[26]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)

In [27]: sparse.coalesce().to_dense()
Out[27]:
tensor([[ 0, 11],
        [ 0,  0]], device='cuda:0', dtype=torch.int16)

In [30]: torch.add(torch.zeros([2,2],dtype=torch.int16, device='cuda'), sparse)
Out[30]:
tensor([[0, 0],
        [0, 0]], device='cuda:0', dtype=torch.int16)
```


[ghstack-poisoned]
